File: 2196-create-binary-tree-from-descriptions/2196-create-binary-tree-from-descriptions.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def createBinaryTree(self, descriptions: List[List[int]]) -> Optional[TreeNode]:
        mp = {}
        root = {}

        for parent, child, isLeft in descriptions:

            if parent not in mp:
                mp[parent] = TreeNode(parent)

            if child not in mp:
                mp[child] = TreeNode(child)

            if isLeft == 1:
                mp[parent].left = mp[child]
            else:
                mp[parent].right = mp[child]

            if root.get(parent, 0) != -1:
                root[parent] = 1

            root[child] = -1

        root_val = 0

        for node, state in root.items():
            if state == 1:
                root_val = node
                break

        return mp[root_val]

File: 2196-create-binary-tree-from-descriptions/2196-create-binary-tree-from-descriptions.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode createBinaryTree(int[][] descriptions) {
        HashMap<Integer, TreeNode> mp = new HashMap<>();
        HashMap<Integer, Integer> root = new HashMap<>();

        for (int[] d : descriptions) {

            int parent = d[0];
            int child = d[1];
            int isLeft = d[2];

            if (!mp.containsKey(parent)) {
                mp.put(parent, new TreeNode(parent));
            }

            if (!mp.containsKey(child)) {
                mp.put(child, new TreeNode(child));
            }

            if (isLeft == 1) {
                mp.get(parent).left = mp.get(child);
            } else {
                mp.get(parent).right = mp.get(child);
            }

            if (root.getOrDefault(parent, 0) != -1) {
                root.put(parent, 1);
            }

            root.put(child, -1);
        }

        int rootVal = 0;

        for (var entry : root.entrySet()) {
            if (entry.getValue() == 1) {
                rootVal = entry.getKey();
                break;
            }
        }

        return mp.get(rootVal);
    }
}